src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkCustomCommandLine.java

```diff
@@ -336,44 +333,52 @@ private int preprocess(
       }
       List<Object> expandedValues =
           maybeExpandDirectories(artifactExpander, originalValues, pathMapper);
-      List<String> stringValues;
+      List<Object /* String | DerivedArtifact */> values;
       if (mapEach != null) {
-        stringValues = new ArrayList<>(expandedValues.size());
+        values = new ArrayList<>(expandedValues.size());
         applyMapEach(
             mapEach,
             expandedValues,
-            stringValues::add,
+            values::add,
             location,
             artifactExpander,
             pathMapper,
             starlarkSemantics);
       } else {
         int count = expandedValues.size();
-        stringValues = new ArrayList<>(expandedValues.size());
+        values = new ArrayList<>(expandedValues.size());
         for (int i = 0; i < count; ++i) {
-          stringValues.add(expandToCommandLine(expandedValues.get(i), pathMapper, mainRepoMapping));
+          values.add(expandToCommandLine(expandedValues.get(i), mainRepoMapping));
         }
       }
       // It's safe to uniquify at this stage, any transformations after this
       // will ensure continued uniqueness of the values
       if ((features & UNIQUIFY) != 0) {
-        int count = stringValues.size();
+        int count = values.size();
         HashSet<String> seen = Sets.newHashSetWithExpectedSize(count);
         int addIndex = 0;
         for (int i = 0; i < count; ++i) {
-          String val = stringValues.get(i);
-          if (seen.add(val)) {
-            stringValues.set(addIndex++, val);
+          Object /* String | DerivedArtifact */ val = values.get(i);
+          // If the path mapper is a no-op, an artifact behaves just like its (trivially mapped)
+          // exec path string.If the path mapper is not a no-op, mapped paths are always distinct
+          // from unmapped paths. We can thus uniquify based on the mapped exec path string in each
+          // case.
+          String valToDeduplicateBy =
```

Comment:
Why deduplicate based on mapped strings instead of unmapped exec paths? Are you just trying to preserve existing behavior, or is it ever expected that two unique artifacts map to the same path (or even weirder, an artifact maps to a path that matches a raw string argument)?

Reply:
* A path mapper could use a hash of the content of a file as its config segment, which could then validly map two different artifacts to the same path. We aren't using this kind of path mapping yet, but it was one of the schemes proposed in the original design doc.
* An artifact with a given exec path behaves differently than its exec path string as a raw string argument, the former is mapped, the latter isn't. See the `uniquifyPathMapped` test for an example of this. This is actually somewhat important as it allows you to create an action that "unmaps" paths, e.g. in jdeps files.
* It is true that an artifact can never (non-trivially) map to a path that matches a raw string argument.

Reply:
Didn't know about the content-based mapping case. Thanks.